### DISCUSSION TITLE
Request for PDF Layer Color Space Information and Document Color Space Detection in PyMuPDF

### DISCUSSION THREAD

**Discussion by Danishkhilji** [pymupdf/PyMuPDF#3829] [Looking for help]
**Overview**

I am working with PyMuPDF and seeking a way to retrieve detailed color space information for individual layers within a PDF. Currently, using doc.get_ocgs(), I can access basic details about each layer (e.g., name, visibility status, usage) but not the color space information.

**Current Approach**::

I am working with PyMuPDF and seeking a way to retrieve detailed color space information for individual layers within a PDF. Currently, using doc.get_ocgs(), I can access basic details about each layer (e.g., name, visibility status, usage) but not the color space information.

**Current Approach**

I have accessed layer information with the following code:
`layers = doc.get_ocgs()
for layer_item in layers.items():
    print(f"Layer: {layer_item}")`
  
 The output provides details like:
 
 `Layer Information:
Layer: (10, {'name': 'Artwork', 'intent': ['View', 'Design'], 'on': True, 'usage': 'Artwork'})
Layer: (7, {'name': 'Keylines', 'intent': ['View', 'Design'], 'on': False, 'usage': 'Artwork'})
Layer: (8, {'name': 'White', 'intent': ['View', 'Design'], 'on': False, 'usage': 'Artwork'})
Layer: (9, {'name': 'Guideline', 'intent': ['View', 'Design'], 'on': False, 'usage': 'Artwork'})
`

However, there is no information about the color space for these layers.

**Objective**

Retrieve Layer Color Space: Is there a way to directly get color space information for individual layers using PyMuPDF?
Document Color Space Detection: If direct retrieval is not possible, can we infer the document’s color space based on layer visibility? For example, could we analyze the document by turning off all layers and examining the resulting color information?

**Additional Details**

I am specifically interested in knowing whether a layer uses CMYK, RGB, or another color space.
Any potential workarounds or alternative methods for detecting color space based on layer visibility would be appreciated.

**Comment by robinwatts** [comment]
What PDF informally calls "Layers" are really "Optional Content Groups".

Optional Content Groups don't have color information associated with them.

Yes, it may be that everything in a given Optional Content group might happen to use the same color model, but there is nothing requiring that. It's perfectly possible to have content in a group that comes from multiple different sources, each with a different color model.

So I'm not sure exactly what you'd want us to return.

**Reply by Danishkhilji** [reply]
@robinwatts Thank you for clarifying that Optional Content Groups (OCGs) don't have color information directly associated with them. That really helps!

As a follow-up, my goal is to ensure that none of the objects within any layer (OCG) contain RGB color, and all should use CMYK. Is there any way I can verify that each object within an OCG is using CMYK, or detect if any RGB color space is present in the layer content?

Ultimately, I need to ensure that my document does not contain any RGB color at all
